python/src/nanoarrow/ipc.py

```diff
@@ -236,9 +243,198 @@ def __repr__(self) -> str:
             return f"<{class_label} <invalid>>"
 
 
+class Writer:
+    def __init__(self):
+        self._writer = None
+        self._desc = None
+        self._iterator = None
+
+    def _is_valid(self) -> bool:
+        return self._writer is not None and self._writer.is_valid()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.close()
+
+    def release(self):
+        """Close stream without writing the end-of-stream marker"""
+        if not self._is_valid():
+            return
+
+        self._writer.release()
+        self._writer = None
+
+    def close(self):
+        """Close stream and write end-of-stream marker"""
+        if not self._is_valid():
+            return
+
+        self._writer.write_end_of_stream()
+        self.release()
+
+    def write_stream(self, obj, schema=None, *, write_schema=None):
```

Comment:
There could in theory also be a `write_array` (or maybe rather `write_batch`) method that takes a single array to write to the stream?